### DISCUSSION TITLE
Follow RFC for self-contained JWT access token

### DISCUSSION THREAD

**Discussion by tnorimat** [keycloak/keycloak#8646] [Ideas]
RFC for self-contained JWT access token has been published.

[RFC 9068 JSON Web Token (JWT) Profile for OAuth 2.0 Access Token](https://datatracker.ietf.org/doc/html/rfc9068)

The current keycloak's access token (and refresh token) is self-contained JWT access token.
It might be good for keycloak's access token to follow this RFC. WDYT?

**Comment by stianst** [comment]
Do you know what the state of it is beyond "proposed standard"?

**Reply by tnorimat** [reply]
The next state is Internet Standard.

The standardization process of RFC in standard track is described by [RFC 2026 The Internet Standards Process -- Revision 3](https://datatracker.ietf.org/doc/html/rfc2026#section-4.1) . It states that there are three maturity level, Proposed Standard, Draft Standard and Internet Standard at the highest maturity level.

Draft Standard is merged into Proposed Standard by [RFC 6410 Reducing the Standards Track to Two Maturity Levels](https://datatracker.ietf.org/doc/html/rfc6410#section-2.3) . The current characteristics of Proposed Standard is described by [RFC 7127  Characterization of Proposed Standards](https://datatracker.ietf.org/doc/html/rfc7127) .

Even if [RFC 6749 The OAuth 2.0 Authorization Framework](https://datatracker.ietf.org/doc/html/rfc6749), its state is Proposed Standard.

**Comment by stianst** [comment]
If this specification gets real traction and not push-back I'm all for following this RFC, but would like to make sure it is on the road to becoming an official standard.

Wondering also a bit on how we could do it, and how much it conflicts with existing claims we have. A basic approach could just be to have a client-scope for it, which is for example what we did for the Microprofile JWT standard.

**Reply by tnorimat** [reply]
I've not yet investigate the difference between the current keycloak's access token and this RFC completely, but I've found the following points.

- JOSE header's "typ" field : [RFC](https://datatracker.ietf.org/doc/html/rfc9068#section-2.1) states that should be "application/at+jwt" or "at+jwt". The latter is recommended. The current keycloak access token set "JWT" to it.
- The fields for meaning group : [RFC](https://datatracker.ietf.org/doc/html/rfc9068#section-2.2.3.1) states that such the field's name  should be "groups".
- The fields for meaning role: [RFC](https://datatracker.ietf.org/doc/html/rfc9068#section-2.2.3.1) states that such the field's name  should be "roles".

**Reply by stianst** [reply]
Quite a few breaking changes, which obviously makes it harder to handle. The groups/roles can be solved by having different "RFC" compliant client scopes, while the typ may be harder.

**Reply by jbman** [reply]
Note: [JSON Web Token RFC 5.2](https://datatracker.ietf.org/doc/html/rfc7519#section-5.1)  states that default "typ" of a JSON web token should be "JWT", but Keycloak sets "Bearer", which is only meant to be used in the JSON response but not inside the JWT.
See https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java#L757

**Reply by lmm-git** [reply]
The RFC does also state that a `client_id` attribute is required. As far as I know it is not possible to add such an attribute to all clients beside of configuring a hardcoded claim for each client.

**Comment by tnorimat** [comment]
@stianst I see, IMO, we need not to follow this specification in a hurry. If it is needed in the future, we would resume this discussion.

**Comment by lmm-git** [comment]
Just stumbled on this discussion as we are currently looking for a standardized way to interpret access tokens and RFC 9068 seems to be the only standard documenting a layout, some attributes and verification steps so far. So maybe we could get the adoption going again?

**Comment by jstale** [comment]
Hi @stianst, is this now officially supported in v25? I saw a change merged that addresses access token type but didn't notice RFC mentioned in release notes.

**Comment by laurids** [comment]
Hi @tnorimat and @stianst.

According to [rfc9068#section-2.1](https://datatracker.ietf.org/doc/html/rfc9068#section-2.1):
> JWT access tokens MUST include this media type in the "typ" header parameter

It then says:
> it is RECOMMENDED that the "application/" prefix be omitted. Therefore, the "typ" value used SHOULD be "at+jwt".

Therefore, it is not just a recommendation, but a requirement that the `typ` header is either `application/at+jwt` or `at+jwt`.
From these two choices, the recommended one is `at+jwt`.

There is already a switch in [DefaultTokenManager.java#L239](https://github.com/keycloak/keycloak/blob/25.0.4/services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java#L239) on the token category to set a specific type for logout tokens.

Could we extend this with a case for the `ACCESS` category? There is already a constant defined for `at+jwt` in [TokenUtil.java#L46](https://github.com/keycloak/keycloak/blob/25.0.4/core/src/main/java/org/keycloak/util/TokenUtil.java#L46).

I realize that this is a breaking change, but without such a change, Keycloak is not compatible with OAuth2 resource servers that verify the `typ` header according to rfc9068.
In case the `ACCESS` category is used for other types of tokens than OAuth2, is there any way to only do this for OAuth2 tokens? I would argue that such a solution could be done in a patch version, and not necessarily require a major version bump, but I understand if you think it should still be done in a major version.
I am looking forward to hearing your thoughts.

**Reply by NobodysNightmare** [reply]
> I realize that this is a breaking change, but without such a change, Keycloak is not compatible with OAuth2 resource servers that verify the typ header according to rfc9068.

This is also what I see as a problem. Right now I wouldn't suggest any resource server to perform the `typ` validation, because IDP support for the new header is not yet there. If now Keycloak is waiting for adoption on the client side, we run into a chicken-and-egg problem.

It would be cool if Keycloak could at least be _configured_ for compliance with the new standard.

**Reply by laurids** [reply]
Hi @NobodysNightmare, I created an issue for this here: https://github.com/keycloak/keycloak/issues/36696
I'm also suggesting to add a configuration option like it has already been done for other types of token.
I would appreciate if you upvoted the issue and maybe added a comment about your perspective as well:)

**Reply by jstale** [reply]
We opted to fork Keycloak for the time being and implement the change in DefaultTokenManager:

```
private String type(TokenCategory category) {
        switch (category) {
            case LOGOUT:
                return TokenUtil.TOKEN_TYPE_JWT_LOGOUT_TOKEN;

            case ACCESS:
                return TokenUtil.TOKEN_TYPE_JWT_ACCESS_TOKEN;
            default:
                return "JWT";
        }
    }
```

This is not ideal but the change is minor and could be a solution for types of projects where at+jwt validation is mandatory.

**Reply by laurids** [reply]
@jstale From version 26.2, there will be an option to make the header type "at+jwt" via a per-client configuration. See https://github.com/keycloak/keycloak/pull/37573

**Reply by jstale** [reply]
> @jstale From version 26.2, there will be an option to make the header type "at+jwt" via a per-client configuration. See #37573

That's great news! Thanks a lot!
